### ISSUE TITLE
A way to quickly select/move between layers

### ISSUE THREAD

**Issue by xerosugar** [mapeditor/tiled#842]
The way it works in Photoshop is pretty good imo and I'd like to see this in Tiled:
- right-click: menu appears, containing all layers that has filled pixels under mouse cursor
- ctrl + right-click: quick-select the layer which has filled pixels under the mouse cursor

**Comment by bjorn** [comment]
Are special tools involved in Photoshop for this? Just asking cause in Tiled, what right-click does depends on the tool. For the default "Stamp Brush" tool, right-click is used to capture an area from the map to paint it elsewhere.

Also the Ctrl and Shift modifiers are already taken on the "Stamp Brush" for drawing lines and circles.

Yet, I can see how these features would be useful, so I wonder if they could be implemented some other way. They could be features of the Rectangular Select tool, which currently still does nothing on right-click. But this may be a little too hidden or hard to access?

**Comment by xerosugar** [comment]
No tools involved in PS. Pressing the Ctrl key, no matter which tool you're using, changes the tool temporarily to the Move tool, allowing you to move a layer or selection. Holding Ctrl and clicking anywhere in the document changes the layer depending on which layer that pixel or whatever it is, is on.

Right click in PS also depends on the tool. With Ctrl pressed when right-clicking (Move tool), you get a menu where you can choose which layer you want to set to the active one. This list contains all layers that has a filled fixel right under the mouse pointer.

Holding Shift in PS also locks the direction. Clicking somewhere in PS and then Shift-clicking with a brush in PS draws a line to that point. In Tiled, you have to keep Shift pressed the whole time for this functionality, and releasing Shift will cancel the functionality completely; I'd prefer pressing Shift at any time to enable the quick line drawing from the last tile I clicked at (similar to PS).

Another thing, atm you have two different shortcuts for selecting stuff; R for tiles; S for shapes. I have a feeling it'd be easier to just have one key to selecting stuff, no matter the tool. In PS the Marquee tool shortcut it 'M'. So if you'd press M on a shape layer, you select shapes. On a tile-layer, you select tiles.

You could always have different flavors of the marquee tool, like in PS. For example Recangular and Circular. To select these, just press and hold the Marquee button to open the pop-up menu you choose from, or Alt-click the tool button to change instantly, like in PS.
Then let's say the user makes a selection, here you could let the user right-click on the selection to open up a menu that says something like "Fill whole selection with active tile" or "Fill selection outline with active tile", with which you could draw a circle if you wanted...

**Comment by bjorn** [comment]
> Holding Shift in PS also locks the direction. Clicking somewhere in PS and then Shift-clicking with a brush in PS draws a line to that point. In Tiled, you have to keep Shift pressed the whole time for this functionality, and releasing Shift will cancel the functionality completely; I'd prefer pressing Shift at any time to enable the quick line drawing from the last tile I clicked at (similar to PS).

I fully agree and have long planned to improve the line drawing in this way, mainly because I'm also used to it working that way in GIMP. The current way is indeed somewhat inconvenient.

> Another thing, atm you have two different shortcuts for selecting stuff; R for tiles; S for shapes. I have a feeling it'd be easier to just have one key to selecting stuff, no matter the tool. In PS the Marquee tool shortcut it 'M'. So if you'd press M on a shape layer, you select shapes. On a tile-layer, you select tiles.

Yeah, I would like to unify these tools at some point, but until then it would already make sense for them to share the same shortcut. The unification would be because often when you want to move stuff, you want to move both the tiles and the objects placed on top. At the moment this kind of operation can get really tedious.

About the right-clicking and the switch to a "move" tool on Ctrl, I'll have to consider how to bring such convenience to Tiled. It does not at the moment seem to me that moving things is so important that it would require Ctrl to be dedicated to it. Yet, of course it is nice that it brings quick access to a context menu with common actions along with it.
